draft-vkrasnov-h2-compression-dictionaries.md

```diff
@@ -138,58 +182,87 @@ The SET_COMPRESSION_CONTEXT frame (type=0xTBA).
 
 ### The SET_DICTIONARY Frame
 
-The SET_DICTIONARY frame (type=0xTBA).
+The SET_DICTIONARY frame (type=0xTBA) contains one to many Dictionary-Entry.
 
 ~~~~
-+-------------+-------------+
-| Dict ID (8) |   Size (8)  |
-+-------------+-------------+
++---------------+---------------+
+|   Dictionary-Entry (+)    ...
++---------------+---------------+
 ~~~~
 
-The SET_DICTIONARY frame can be sent from the server to the client, on any client
-initiated stream in the open or half-closed (remote) states. The SET_DICTIONARY
-frame MUST precede any DATA frames on that stream. The SET_DICTIONARY frame SHOULD
-be followed by sufficient DATA frames to fill Size octets after decompression,
-even if an RST frame was received for that stream. If not enough DATA was sent,
-the Dictionary for the given ID is considered uninitialized.
+A Dictionary-Entry field is encoded as follows:
+
+~~~~
++-------------------------------+
+|       Dictionary-ID (8)       |
++---+---------------------------+
+| P |        Size (7+)          |
++---+---------------------------+
+| E?|      Truncate? (7+)       |
++---+---------------------------+
+|           Offset? (8+)        |
++-------------------------------+
 
-More than one SET_DICTIONARY frames MAY be sent on a given stream.
+~~~~
 
-The SET_DICTIONARY frame contains the following fields:
+The SET_DICTIONARY frame can be sent from the server to the client, on any
+client initiated stream in the open or half-closed (remote) states. The
+SET_DICTIONARY frame MUST precede any DATA frames on that stream. The
+SET_DICTIONARY frame SHOULD be followed by sufficient DATA frames to build the
+dictionaries.
+If a RST frame was received for the stream before sufficient DATA was sent, the
+dictionaries are reset.
 
-Dict ID:
+The Dictionary-Entry contains the following fields:
+
+Dictionary-ID:
 : an 8-bit ID, indicates the dictionary. MUST be lower than the value agreed by
-the SETTINGS_COMPRESSION_SETTINGS setting.
+the SETTINGS_COMPRESSION setting.
 
 Size:
-: Indicates how many octets (as a power of 2) of the given stream will be used.
-If Size is greater than the length of the transmitted data, then all of the data
-will be used.
-
-The SET_DICTIONARY frame defines the following flag:
+: Indicates how many octets of the stream will be used for the dictionary.
+Size is represented as an integer with 7-bit prefix (see {{RFC1951}},
+Section 5.1). If P is set, the actual number of octets to use is 2 to the power
+of Size. If the computed value is greater than the length of the decompressed
+DATA, use all the available DATA.
+
+Truncate:
+: An optional field, represented as an integer with 7-bit prefix. Present when
+the APPEND flag is set. Truncate indicates the number of octets to keep of the
+existing dictionary, before appending the new data to it. If E is set, then
+Truncate is ignored, and new data is appended at the end. If Truncate is zero,
+then the dictionary is replaced, as if APPEND was unset.
```

Comment:
which part of the dictionary are we keeping? I'd prefer we'd keep the end of the current dictionary, as it's more likely to be relevant, but as written it's not really defined.

Reply:
truncate removes the tail, the head of the dictionary is implicitly removed once it overflows.
unless there is any other reason to explicitly remove the head?

`d.dict = tail(append(old_dict_data, new_dict_data), 1 << settings.DSize)`

Reply:
Perhaps we can add another bit to indicate if the truncation is a `head` or a `tail`?
Something like:
```
+---+---+-----------------------+
| E?| D?|    Truncate? (6+)     |
+---+---+-----------------------+
```

D (stands for direction), if D == 1 keep the last Truncate octets, otherwise keep the first Trucate octets.

Reply:
Hmm, I didn't understand that we're treating the dictionary as a sliding window and dropping its start when more recent data comes in and the dictionary overflows. In that light, I'm not sure there's real use-case for the "truncate" part.

Reply:
Yeah, it is a sliding window.
Still, truncate can give you some interesting options. I think we should keep it.